shell/renderer/api/atom_api_context_bridge.cc

```diff
@@ -0,0 +1,426 @@
+    for (auto& tup : functions_) {
+      std::get<0>(tup).Reset();
+      std::get<1>(tup).Reset();
+    }
+    reverse_binding_store_.Reset();
+    delete this;
+  }
+
+  std::vector<FunctionContextPair>& functions() { return functions_; }
+
+ private:
+  // { function, owning_context }[]
+  std::vector<FunctionContextPair> functions_;
+  v8::Global<v8::Object> reverse_binding_store_;
+};
+
+std::map<content::RenderFrame*, RenderFramePersistenceStore*> store_map_;
+
+RenderFramePersistenceStore* GetOrCreateStore(
+    content::RenderFrame* render_frame) {
+  if (store_map_.find(render_frame) == store_map_.end()) {
+    store_map_[render_frame] = new RenderFramePersistenceStore(render_frame);
+  }
+  return store_map_[render_frame];
+}
+
+// Sourced from "extensions/renderer/v8_schema_registry.cc"
+// Recursively freezes every v8 object on |object|.
+void DeepFreeze(const v8::Local<v8::Object>& object,
+                const v8::Local<v8::Context>& context) {
+  v8::Local<v8::Array> property_names =
+      object->GetOwnPropertyNames(context).ToLocalChecked();
+  for (uint32_t i = 0; i < property_names->Length(); ++i) {
+    v8::Local<v8::Value> child =
+        object->Get(context, property_names->Get(context, i).ToLocalChecked())
+            .ToLocalChecked();
+    if (child->IsObject())
+      DeepFreeze(v8::Local<v8::Object>::Cast(child), context);
+  }
+  object->SetIntegrityLevel(context, v8::IntegrityLevel::kFrozen);
+}
+
+// This is unsafe as it is a hack around move semantics and promises requiring
+// two callbacks We use this class to encapsulate the promise and retrieve it
+// from whichever handler is called It is guarunteed that only one handler is
+// called so the promise is moved to the correct handler once it is called and
+// destroyed afterwards.
+class UnsafePromiseHandle {
+ public:
+  explicit UnsafePromiseHandle(
+      electron::util::Promise<v8::Local<v8::Value>> promise)
+      : promise_(std::move(promise)), has_handle_(true) {}
+
+  electron::util::Promise<v8::Local<v8::Value>> TakePromise() {
+    DCHECK(has_handle_);
+    has_handle_ = false;
+    return std::move(promise_);
+  }
+
+ private:
+  electron::util::Promise<v8::Local<v8::Value>> promise_;
+  bool has_handle_ = false;
+};
+
+}  // namespace
+
+v8::Local<v8::Value> PassValueToOtherContext(v8::Local<v8::Context> source,
+                                             v8::Local<v8::Context> destination,
+                                             v8::Local<v8::Value> value) {
+  // Proxy promises as they have a safe and guarunteed memory lifecycle (unlike
+  // functions)
+  if (value->IsPromise()) {
+    v8::Context::Scope destination_scope(destination);
+
+    auto v8_promise = v8::Local<v8::Promise>::Cast(value);
+    util::Promise<v8::Local<v8::Value>> promise(destination->GetIsolate());
+    v8::Local<v8::Promise> handle = promise.GetHandle();
+    UnsafePromiseHandle* p_handle = new UnsafePromiseHandle(std::move(promise));
```

Comment:
The `util::Promise` can be allocated on heap, so `UnsafePromiseHandle` should not be needed:

```c++
auto* promise = new util::Promise<v8::Local<v8::Value>>(isolate);
auto then_cb = base::BindOnce([](util::Promise* promise) {
  promise->Resolve();
  delete promise;
}, promise);
```